//
// Copyright 2022 Ant Group Co., Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

syntax = "proto3";

package spu.psi.proto;

message SealParamsProto {
  uint32 poly_modulus_degree = 1;
  uint32 plain_modulus = 2;
  repeated uint64 coeff_modulus = 3;
}

message LabelPsiParamsProto {
  uint32 hash_func_count = 1;
  uint32 table_size = 2;
  uint32 max_items_per_bin = 3;

  uint32 felts_per_item = 4;

  uint32 ps_low_degree = 5;
  repeated uint32 query_powers = 6;

  SealParamsProto seal_params = 7;
}

message OprfProto {
  repeated bytes data = 1;
}

message EncryptedPowersProto {
  uint64 power = 1;
  repeated bytes ciphertexts = 2;
}

message QueryRequestProto {
  bytes relin_keys = 1;
  repeated EncryptedPowersProto encrypted_powers = 2;
}

message QueryResultProto {
  uint32 bundle_idx = 1;
  bytes ciphertext = 2;
  uint32 label_byte_count = 3;
  uint32 nonce_byte_count = 4;
  repeated bytes label_results = 5;
}

message QueryResponseProto {
  repeated QueryResultProto results = 1;
}
